examples/go/tagging/gotag-fn/handler.go

```diff
@@ -0,0 +1,177 @@
+package function
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+
+	handler "github.com/openfaas-incubator/go-function-sdk"
+	"github.com/pelletier/go-toml"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/vapi/rest"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// tagInfo is part of the toml vcconfig file
+type tagInfo struct {
+	Action string
+	Urn    string
+}
+
+// VCenter is part of the toml vcconfig file
+type vCenterAuth struct {
+	Password string
+	Server   string
+	User     string
+}
+
+// Config represents the toml vcconfig file
+type vcConfig struct {
+	Tag     tagInfo
+	VCenter vCenterAuth
+}
+
+// The request from vmware-event-router.
+type dataObj struct {
+	Data nest1 `json:"data"`
+}
+type nest1 struct {
+	Vm nest2 `json:"Vm"`
+}
+type nest2 struct {
+	Vm vmInfo `json:"Vm"`
+}
+type vmInfo struct {
+	Value string `json:"Value"`
+	Type  string `json:"Type"`
+}
+
+// Handle a function invocation
+func Handle(req handler.Request) (handler.Response, error) {
+	ctx := context.Background()
+
+	// Retrieve vCenter and tag info from the config file.
+	vcc := loadTomlCfg()
+
+	// Incoming request object.
+	var mor dataObj
+	err := json.Unmarshal(req.Body, &mor)
+	if err != nil {
+		return handler.Response{
+			Body:       nil,
+			StatusCode: http.StatusBadRequest,
+		}, err
+	}
+
+	if mor.Data.Vm.Vm.Value == "" || mor.Data.Vm.Vm.Type == "" {
+		return handler.Response{
+			Body:       nil,
+			StatusCode: http.StatusBadRequest,
+		}, errors.New("managed Reference Object information must not be empty")
+	}
+
+	// Fill information in the request into a govmomi type.
+	var moRef types.ManagedObjectReference
+	moRef.Value = mor.Data.Vm.Vm.Value
+	moRef.Type = mor.Data.Vm.Vm.Type
+
+	// Connect to vcenter with govmomi client
+	log.Println("Connecting to vCenter")
+	clients := connect(ctx, vcc.VCenter)
+	log.Println("Connection successful.")
+
+	// tag the VM moRef with tagID (URN)
+	clients.tagVM(ctx, moRef, vcc.Tag.Urn)
+
+	// Respond with a message that tagging of vm was complete
+	message := fmt.Sprintf(
+		"%v was tagged with %v",
+		moRef.Value,
+		vcc.Tag.Urn,
+	)
+	log.Println(string(message))
+
+	err = clients.Govmomi.Logout(ctx)
+	if err != nil {
+		log.Printf("Unable to log out of govmomi due to %v.", err)
+	}
+	err = clients.Rest.Logout(ctx)
+	if err != nil {
+		log.Printf("Unable to log out of rest due to %v.", err)
+	}
+	log.Println("Logged out of Govmomi and Rest clients.")
```

Comment:
this log statement could be wrong if logout from vc succeeds, but not from the tagging API (not atomic)

Reply:
I got rid of the log statement completely. Not sure if it is useful to anyone.